src/inferencer/constraintStore.ts

```diff
@@ -1,11 +1,90 @@
+import { Type, Variable, Primitive } from '../types'
+
 export const constraintStore = new Map()
 
-export function updateTypeConstraints(newConstraintLhs: number, newConstraintRhs: number | string) {
-  // step 3a. Add new constraint to constraintStore map
-  // e.g. T1 = T2, T2 = number
-  constraintStore.set(newConstraintLhs, newConstraintRhs)
+export function updateTypeConstraints(newConstraintLhs: Type, newConstraintRhs: Type) {
+  // console.log('updateTypeConstraints')
+  // console.log(newConstraintLhs + ' ' + newConstraintRhs)
+  solveConstraint(newConstraintLhs, newConstraintRhs)
+  // if (detectError(constraintStore)) console.log('Error! Rule 9')
+}
+
+function solveConstraint(constraintLhs: Type, constraintRhs: Type) {
+  // check if both key and value are not base types (Rule 1)
+  if (!is_base_type(constraintLhs) && !is_base_type(constraintRhs)) {
+    // check if key is not a type variable and value is a type variable (Rule 2)
+    if (is_base_type(constraintLhs) && is_type_variable(constraintRhs)) {
```

Comment:
I don't think we shld be checking `is_base_type(constraintLhs)`. What if constraintLhs is function type?
Suggestion: `!is_type_variable(constraintLhs)`

Reply:
updated